route/route-free-trial-request.js

```diff
@@ -1,13 +1,119 @@
 const bodyParse = require('body-parser').json();
 const Student = require('./../schemas/student');
+const Teacher = require('./../schemas/teacher');
+const makeAvailability = require('./../lib/makeAvailabilityObject');
+const availableTeachers = require('./../lib/filterTeachersByAvailability');
+const createAddress = require('./../lib/createAddrString');
+const findDistance = require('./../lib/distanceMatrixAPI');
+const createTimeSlot = require('./../lib/createTimeSlot');
 
 module.exports = router => {
   router.post('*api/free-trial-request', bodyParse, (req, res) => {
+    const postResponse = {};
     const bodyParams = Object.keys(req.body);
-    // Need to validate that bodyParams contains all required params
+
     if (!bodyParams.length) {
-      return res.status(400).send('Bad Request body');
+      return res.status(400).send('Bad Request: Request must include body');
+    }
+
+    const {
+      addressOne,
+      city,
+      state,
+      zip,
+      instrument,
+      musicStyle,
+      allergies,
+      specialNeeds,
+      comments,
+      experienceLevel,
+      availability,
+    } = req.body;
+
+    if (
+      !(
+        addressOne &&
+        city &&
+        state &&
+        zip &&
+        instrument &&
+        musicStyle &&
+        allergies &&
+        specialNeeds &&
+        comments &&
+        experienceLevel &&
+        availability &&
+        'hasInstrument' in req.body
+      )
+    ) {
+      return res
+        .status(400)
+        .send('Bad Request: Request body missing required properties');
     }
-    return new Student(req.body).save().then(student => res.json(student));
+
+    const studentAvailability = makeAvailability(availability);
+    let filteredTeachersArray = [];
+
+    return new Student(req.body)
```

Comment:
This is a great first formulation of the algorithm! I have a possible refactor method in mind that will make it more readable and avoid the long string of chained `.then`'s. Basically, every time there's a `.then`, that means the previous operation is returning a Promise, which means the `.then` can be replaced with `async/await`. We should discuss more, either this Saturday in person or one night this week on a call.